File: pet-api/internal/queries/owner_queries.go
package queries

import (
	"database/sql"
	"github.com/google/uuid"
	"pet-api/internal/database"
	"time"
)

type OwnerQueries struct {
	BaseQueries
}

type Owner struct {
	ID        uuid.UUID
	Name      string
	Email     string
	Phone     string
	Address   string
	CreatedAt time.Time
	UpdatedAt time.Time
}

func NewOwnerQueries(db database.Service) OwnerQueries {
	baseQueries := NewBaseQueries(db)
	return OwnerQueries{baseQueries}
}
func (oq *OwnerQueries) QueryRow(query string, args ...interface{}) *sql.Row {
	return oq.QueryRow(query, args...)
}

func (oq *OwnerQueries) Create(owner Owner) error {
	owner.ID = uuid.New()
	query := `INSERT INTO owners (owner_id, owner_name, owner_email, owner_phone, owner_address) VALUES (?, ?, ?, ?, ?)`
	_, err := oq.Query(query, owner.ID, owner.Name, owner.Email, owner.Phone, owner.Address)
	return err
}

func (oq *OwnerQueries) Update(owner Owner) error {
	query := `UPDATE owners SET owner_name=?, owner_email=?, owner_phone=?, owner_address=? WHERE owner_id=?`
	_, err := oq.Query(query, owner.Name, owner.Email, owner.Phone, owner.Address, owner.ID)
	return err
}

func (oq *OwnerQueries) Delete(id uuid.UUID) error {
	query := `DELETE FROM owners WHERE owner_id=?`
	_, err := oq.Query(query, id)
	return err
}

func (oq *OwnerQueries) GetAll() ([]Owner, error) {
	query := `SELECT * FROM owners`
	rows, err := oq.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var owners []Owner
	for rows.Next() {
		var owner Owner
		err := rows.Scan(&owner.ID, &owner.Name, &owner.Email, &owner.Phone, &owner.Address, &owner.CreatedAt, &owner.UpdatedAt)
		if err != nil {
			return nil, err
		}
		owners = append(owners, owner)
	}

	if err = rows.Err(); err != nil {
		return nil, err
	}

	return owners, nil
}

func (oq *OwnerQueries) GetById(id uuid.UUID) (*Owner, error) {
	query := `SELECT * FROM owners WHERE owner_id = ?`
	row := oq.QueryRow(query, id)

	var owner Owner
	err := row.Scan(&owner.ID, &owner.Name, &owner.Email, &owner.Phone, &owner.Address, &owner.CreatedAt, &owner.UpdatedAt)
	if err == sql.ErrNoRows {
		return nil, nil
	} else if err != nil {
		return nil, err
	}

	return &owner, nil
}

File: pet-api/src/db/queries/owner_queries.rs
use std::sync::Arc;

use sqlx::Row;

use crate::models::{
    owner_model::{OwnerModel, OwnerWithPets},
    pet_model::PetModel,
};

pub struct OwnerQueries {
    db: Arc<sqlx::MySqlPool>,
}

impl OwnerQueries {
    pub fn new(db: Arc<sqlx::MySqlPool>) -> Self {
        Self { db }
    }

    pub async fn insert_owner(
        &self,
        owner_id: String,
        owner_name: String,
        owner_email: String,
        owner_phone_number: String,
        owner_address: String,
    ) -> Result<u64, sqlx::Error> {
        sqlx::query(r#"INSERT INTO owner (owner_id, owner_name, owner_email, owner_phone_number, owner_address) VALUES (?, ?, ?, ?, ?)"#)
            .bind(owner_id)
            .bind(owner_name)
            .bind(owner_email)
            .bind(owner_phone_number)
            .bind(owner_address)
            .execute(&*self.db)
            .await
            .map(|done| done.rows_affected())
    }

    pub async fn select_owner(&self, owner_id: String) -> Result<OwnerModel, sqlx::Error> {
        sqlx::query_as("SELECT * FROM owner WHERE owner_id = ?")
            .bind(owner_id)
            .fetch_one(&*self.db)
            .await
    }

    pub async fn select_all_owners(
        &self,
        limit: i32,
        offset: i32,
        search: Option<String>,
    ) -> Result<Vec<OwnerModel>, sqlx::Error> {
        let mut query = String::from("SELECT * FROM owner ");

        if let Some(search_term) = search {
            query.push_str("WHERE owner_name LIKE ? ");
            query.push_str("ORDER BY owner_name ");
            query.push_str("LIMIT ? OFFSET ?");

            sqlx::query_as::<_, OwnerModel>(&query)
                .bind(format!("%{}%", search_term))
                .bind(limit)
                .bind(offset)
                .fetch_all(&*self.db)
                .await
        } else {
            query.push_str("ORDER BY owner_name ");
            query.push_str("LIMIT ? OFFSET ?");

            sqlx::query_as::<_, OwnerModel>(&query)
                .bind(limit)
                .bind(offset)
                .fetch_all(&*self.db)
                .await
        }
    }

    pub async fn delete_owner(&self, owner_id: String) -> Result<u64, sqlx::Error> {
        sqlx::query("DELETE FROM owner WHERE owner_id = ?")
            .bind(owner_id)
            .execute(&*self.db)
            .await
            .map(|done| done.rows_affected())
    }
    pub async fn update_owner(
        &self,
        owner_id: String,
        owner_name: Option<String>,
        owner_email: Option<String>,
        owner_phone_number: Option<String>,
        owner_address: Option<String>,
    ) -> Result<u64, sqlx::Error> {
        let mut query_string = String::from("UPDATE owner SET ");
        let mut params = Vec::new();

        if let Some(name) = owner_name {
            query_string.push_str("owner_name = ?, ");
            params.push(name);
        }

        if let Some(email) = owner_email {
            query_string.push_str("owner_email = ?, ");
            params.push(email);
        }

        if let Some(phone_number) = owner_phone_number {
            query_string.push_str("owner_phone_number = ?, ");
            params.push(phone_number);
        }

        if let Some(address) = owner_address {
            query_string.push_str("owner_address = ?, ");
            params.push(address);
        }

        if query_string.ends_with(", ") {
            query_string.truncate(query_string.len() - 2);
        }

        query_string.push_str(" WHERE owner_id = ?");
        params.push(owner_id);

        let mut query = sqlx::query(&query_string);

        for param in params {
            query = query.bind(param);
        }

        let result = query.execute(&*self.db).await?;

        Ok(result.rows_affected())
    }

    pub async fn get_owner_and_pets(&self, owner_id: String) -> Result<OwnerWithPets, sqlx::Error> {
        let owner = match self.select_owner(owner_id.clone()).await {
            Ok(owner) => owner,
            Err(_) => return Err(sqlx::Error::RowNotFound),
        };

        let pets =
            sqlx::query_as::<_, PetModel>("SELECT * FROM PET WHERE owner_id = ? ORDER BY pet_name")
                .bind(owner_id)
                .fetch_all(&*self.db)
                .await?;

        Ok(OwnerWithPets { owner, pets })
    }

    pub async fn count_all_owners(&self, search: Option<String>) -> Result<i64, sqlx::Error> {
        let mut query = String::from("SELECT COUNT(*) as count FROM owner ");

        if let Some(search_term) = search {
            query.push_str(&format!("WHERE owner_name LIKE '%{}%' ", search_term));
        }

        let query = sqlx::query(&query);

        query
            .fetch_one(&*self.db)
            .await
            .map(|row: sqlx::mysql::MySqlRow| row.get("count"))
    }
}
